monkey/infection_monkey/exploit/http_agent_binary_server.py

```diff
@@ -0,0 +1,172 @@
+import logging
+import threading
+from http.server import HTTPServer
+from ipaddress import IPv4Address
+from typing import Callable, Optional, Type
+from uuid import uuid4
+
+from common import OperatingSystem
+from common.types import Event, Lock, NetworkPort
+from common.utils.code_utils import insecure_generate_random_string
+from infection_monkey.network import TCPPortSelector
+from infection_monkey.network.tools import get_interface_to_target
+from infection_monkey.utils.threading import create_daemon_thread
+
+from .agent_binary_request import (
+    AgentBinaryDownloadReservation,
+    AgentBinaryDownloadTicket,
+    AgentBinaryTransform,
+    ReservationID,
+)
+from .http_agent_binary_request_handler import AgentBinaryHTTPRequestHandler
+
+logger = logging.getLogger(__name__)
+
+AgentBinaryHTTPHandlerFactory = Callable[[], Type[AgentBinaryHTTPRequestHandler]]
+
+
+def use_agent_binary(agent_binary: bytes) -> bytes:
+    return agent_binary
+
+
+class HTTPAgentBinaryServer:
+    """
+    Serves Agent binaries over HTTP
+
+    Allows clients to register for an Agent binary to be served. The server will serve the
+    requested binary until it is deregistered or the server is stopped.
+
+    :param tcp_port_selector: The TCP port selector to use
+    :param get_handler_class: A function that returns the HTTP handler class to use
+    :param lock: A lock to use
+    :param poll_interval: The interval to poll for server shutdown, in seconds
+    """
+
+    def __init__(
+        self,
+        tcp_port_selector: TCPPortSelector,
+        get_handler_class: AgentBinaryHTTPHandlerFactory,
+        create_event: Callable[[], Event],
+        lock: Lock,
+        poll_interval: float = 0.5,
+    ):
+        self._tcp_port_selector = tcp_port_selector
+        self._handler_class = get_handler_class()
+        self._create_event = create_event
+        self._lock = lock
+        self._poll_interval = poll_interval
+        self._port: Optional[NetworkPort] = None
+        self._server: Optional[HTTPServer] = None
+        self._server_thread: Optional[threading.Thread] = None
+
+    def register(
+        self,
+        operating_system: OperatingSystem,
+        requestor_ip: IPv4Address,
+        agent_binary_transform: AgentBinaryTransform = use_agent_binary,
+    ) -> AgentBinaryDownloadTicket:
+        """
+        Register to download an Agent binary
+
+        If the server is not running, it will be started.
+
+        :param operating_system: The operating system for the Agent binary to serve
+        :param requestor_ip: The IP address of the client that will download the Agent binary
+        :param agent_binary_transform: A function that transforms the Agent binary before serving.
```

Comment:
I have a vague idea right now that this has something to do with the dropper script, but can we make this more descriptive? Someone reading this for the first time can't make much of it.